lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,468 @@
+  CORSUnsafeRequestHeaderNames() {
+    // TODO
+  }
+}
+
+// https://fetch.spec.whatwg.org/#headers-class
+const kHeaderList = Symbol('headerList');
+const kGuard = Symbol('guard');
+class Headers {
+  // https://fetch.spec.whatwg.org/#dom-headers
+  constructor(init) {
+    this[kHeaderList] = new HeaderList();
+    this[kGuard] = 'none';
+    if (init !== undefined){
+      fillHeaders(this, init);
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-headers-append
+  append(name, value) {
+    appendHeader(this, name, value);
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-headers-delete
+  delete(name) {
+    name = normalizeName(name);
+    const lowerName = StringPrototypeToLowerCase(name);
+    if (this[kGuard] === 'immutable') {
+      throw new TypeError('this Headers object is immutable');
+    }
+    if (this[kGuard] === 'request' && forbiddenHeaderNames.has(lowerName)) {
+      return;
+    }
+    if (this[kGuard] === 'request-no-cors' &&
+        !noCORSSafelistedRequestHeaderNames.has(lowerName) &&
+        !privilegedNoCORSRequestHeaderNames.has(lowerName)) {
+      return;
+    }
+    if (this[kGuard] === 'response' &&
+        forbiddenResponseHeaderNames.has(lowerName)) {
+      return;
+    }
+    if (!this[kHeaderList].contains(lowerName)) {
+      return;
+    }
+    this[kHeaderList].delete(lowerName);
+    if (this[kGuard] === 'request-no-cors') {
+      removePrivilegedNoCORSRequestHeaders(this);
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-headers-get
+  get(name) {
+    name = normalizeName(name);
+    return this[kHeaderList].get(name);
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-headers-has
+  has(name) {
+    name = normalizeName(name);
+    return this[kHeaderList].contains(name);
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-headers-set
+  set(name, value) {
+    name = normalizeName(name);
+    value = normalizeValue(value);
+    const lowerName = StringPrototypeToLowerCase(name);
+    if (this[kGuard] === 'immutable') {
+      throw new TypeError('this Headers object is immutable');
+    }
+    if (this[kGuard] === 'request' && forbiddenHeaderNames.has(lowerName)) {
+      return;
+    }
+    if (this[kGuard] === 'request-no-cors' &&
+        !noCORSSafelistedRequestHeaderNames.has(lowerName)) {
+      return;
+    }
+    if (this[kGuard] === 'response' &&
+        forbiddenResponseHeaderNames.has(lowerName)) {
+      return;
+    }
+    this[kHeaderList].set(name, value);
+    if (this[kGuard] === 'request-no-cors') {
+      removePrivilegedNoCORSRequestHeaders(this);
+    }
+  }
+
+  *entries() {
+    const headers = this[kHeaderList].sortAndCombine();
+    yield* headers;
+  }
+
+  *keys() {
+    for (const entry of this.entries()) {
+      yield entry[0];
+    }
+  }
+
+  *values() {
+    for (const entry of this.entries()) {
+      yield entry[1];
+    }
+  }
+
+  [SymbolIterator]() {
+    return this.entries();
+  }
+
+  forEach(callback) {
```

Comment:
Where does this come from? is this apart of being an `iterable`? If so is there a better way to indicate this such as extending from some parent class?

Reply:
I don't know. This is one of the many things that I don't understand about the spec. I don't know if it's supposed to be here because of `iterable` or something else, but I added the method to make some of the web platform tests pass.